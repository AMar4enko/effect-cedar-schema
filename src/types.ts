export type IdentifiableEntity = { _tag: string, id: string }

export interface SerializedIdentifier {
  entityType: string
  entityId: string
}

export type SerializedType = 
  | { boolean: boolean }
  | { long: number }
  | { string: string }
  | { record: Record<string, SerializedType> }
  | { set: SerializedType[] }
  | { entityIdentifier: SerializedIdentifier }

export type SerializedAttributes = Record<string, SerializedType>