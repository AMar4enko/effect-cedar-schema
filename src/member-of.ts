import * as Schema from "effect/Schema"

export const CedarMemberOf = Symbol(`@cedar-schema/memberOf`)

export const memberOf = <T extends Schema.Struct<Schema.Struct.Fields>[]>(...memberOf: T) => 
  (schema: Schema.Struct<Schema.Struct.Fields>) => Schema.annotations(schema, { [CedarMemberOf]: memberOf.map(({ ast }) => ast)})
  