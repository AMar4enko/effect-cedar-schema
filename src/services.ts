import { GenericTag } from "effect/Context"

export interface CedarSchema {
  readonly defaultNamespace: string
  readonly namespace: Map<string, {
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    entityTypes: Map<string, Record<string, any>>
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    commonTypes: Map<string, Record<string, any>>
    // biome-ignore lint/suspicious/noExplicitAny: <explanation>
    actions: Map<string, Record<string, any>>  
  }>
}

export const CedarSchema = GenericTag<CedarSchema>(`@cedar-schema/schema`)
