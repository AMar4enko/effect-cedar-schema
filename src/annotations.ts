import { Annotated, getAnnotation } from "effect/SchemaAST"
import * as Effect from 'effect/Effect'
import { CedarSchema } from './services.js'

export const CedarNamespace = Symbol(`@cedar-schema/namespace`)

export const getCedarNamespace = (annotated: Annotated) => getAnnotation<string>(annotated, CedarNamespace).pipe(
  Effect.orElse(() => CedarSchema.pipe(Effect.map(({ defaultNamespace }) => defaultNamespace)))
)