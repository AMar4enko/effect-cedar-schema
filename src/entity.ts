import { AST, Schema } from "@effect/schema"
import { Class, Struct, TaggedClass as TaggedClassImpl } from "@effect/schema/Schema"
import { CedarNamespace } from "./annotations.js"
import { Effect, Equal, FiberRef, identity, Match, Record as R } from "effect"
import { UnknownException } from "effect/Cause"
import { SerializedIdentifier, FullSerializedType, SerializedType } from "./types.js"
import { IdentifierAnnotationId, isUndefinedKeyword } from "@effect/schema/AST"
import * as Hash from "effect/Hash"

type EntitySchema = Schema.Schema.All & { [EntityTypeId]: typeof EntityTypeId }

export const EntityTypeId: unique symbol = Symbol(`effect-cedar/EntityTypeId`)

export interface SerializedEntity extends Hash.Hash {
  identifier: SerializedIdentifier
  attributes: Record<string, FullSerializedType>
}

const EntitiesRef = FiberRef.unsafeMake(new Map<string, SerializedEntity>())

export type CompilerFunction = (v: unknown) => Effect.Effect<FullSerializedType, Error, never>

export const getEntities = FiberRef.get(EntitiesRef).pipe(
  Effect.map((map) => [...map.values()])
)

export const makeSerialisedEntity = function (identifier: SerializedIdentifier, attributes: Record<string, FullSerializedType>): SerializedEntity {
  const id = { ...identifier }
  const idHash = Hash.hash(`${identifier.entityType}${identifier.entityId}`)
  Hash.cached(id, idHash)
  const o = {
    identifier: id,
    attributes
  }

  Hash.cached(o, idHash)

  return o as any
}

export type FieldsWithParents< Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never> = 
  [MembersOf] extends [never] 
    ? Fields & { id: typeof Schema.String }
    : Fields & { id: typeof Schema.String; parents: Schema.optional<Schema.Array$<Schema.Union<MembersOf>>> }


export interface TaggedClass<Self, Tag extends string, Fields extends Struct.Fields, Proto> extends
    Class<
      Self,
      Fields,
      Struct.Encoded<Fields>,
      Struct.Context<Fields>,
      Struct.Constructor<Omit<Fields, "_tag">>,
      {},
      Proto
    >
  {
    readonly _tag: Tag
  }

export interface Entity<Self, Tag extends string, Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never> extends 
  TaggedClass<
    Self,
    Tag,
    FieldsWithParents<Fields, MembersOf>,
    {
      serialize: () => Effect.Effect<FullSerializedType, never, never>
    }
  >
  {
    [EntityTypeId]: typeof EntityTypeId
    readonly _tag: Tag
  }

/**
 * Gotta prohibit entity fields to be classes
 */
export type EntityField<F> = F extends new (...args: any[]) => any
  ? never
  : | Schema.Schema.All
    | Schema.PropertySignature.All

export const compile: (ast: AST.AST) => CompilerFunction = Match.type<AST.AST>().pipe(
  Match.when(
      (ast) => ast._tag === `Declaration` && ast.annotations[EntityTypeId] === EntityTypeId,
      () => (value: InstanceType<Entity<any, any, any, any>>) => value.serialize()
  ),
  Match.tag(`TupleType`, (ast) => {
    const runCompile = compile(ast.rest[0].type)

    return (value: any[]): Effect.Effect<SerializedType[], never, never> => Effect.all(value.map(runCompile)) as any
  }),
  Match.tag(`Union`, (ast): CompilerFunction => {
    const hasUndefined = ast.types.some((type) => isUndefinedKeyword(type))
    const withoutUndefined = ast.types.filter((type) => !isUndefinedKeyword(type))
    const map = new Map<string, CompilerFunction>()

    if (withoutUndefined.length === 1) {
      return compile(withoutUndefined[0])
    }

    return (value: unknown) => {
      if (value === undefined && hasUndefined) {
        return Effect.succeed(value)
      }

      /** Union here can only be union of multiple entity types, hence we are safe to assume we can serialize it */
      return ((value as any).serialize as any)()
    }
  }),
  Match.tag(`TypeLiteral`, (ast): any => {
    const compilers = R.fromEntries(
      ast.propertySignatures.map((propSignature) => [String(propSignature.name), compile(propSignature.type)] as const)
    )

    return (value: Record<PropertyKey, any>): Effect.Effect<FullSerializedType, never, never> => Effect.all(R.map(compilers, (runCompile, key) => runCompile(value[key]))) as any
  }),
  Match.orElse(() => Effect.succeed)
)
  
const toApiJSON = (schema: Entity<any, any, any, any>) => {
  const ast = schema.ast as AST.Transformation
   if (ast.to._tag !== `Declaration` || (ast.to.typeParameters.length === 0 || ast.to.typeParameters[0]._tag !== `TypeLiteral`)) { 
    throw new UnknownException(`unexpected AST`)
  } else {}
  const typeLiteral = ast.to.typeParameters[0]

  const NS = ast.to.annotations[CedarNamespace]
  const identifier: Record<`entityType` | `entityId`, (value: any) => string> = {
    entityType: () => `${NS}::${ast.to.annotations[IdentifierAnnotationId]}`,
    entityId: identity
  }
  const attributes: Record<string, CompilerFunction> = {}
  
  typeLiteral.propertySignatures.forEach((propSignature) => {
    if (propSignature.name === `id`) {
      identifier.entityId = (value: any) => value[`id`]
    } else if (propSignature.name !== `_tag` && !isUndefinedKeyword(propSignature.type)) {
      attributes[String(propSignature.name)] = compile(propSignature.type)
    }
  })
  
  return (value: Record<PropertyKey, any>): Effect.Effect<FullSerializedType, never, never> => Effect.gen(function* () {
    const a = yield* FiberRef.get(EntitiesRef)

    const key = `${identifier.entityType(value)}_${identifier.entityId(value)}`

    return yield* Effect.fromNullable(a.get(key))
      .pipe(
        Effect.orElse(() => 
          Effect.all(R.map(attributes, (runCompile, key) => runCompile(value[key]))).pipe(
            Effect.map((attributes) => 
              makeSerialisedEntity(
                R.map(identifier, (a) => a(value)),
                attributes
              )
            )
          ).pipe(
            Effect.tap((value) => FiberRef.update(EntitiesRef, (a) => a.set(key, value)))
          )
        ),
        Effect.map(({ identifier }) => identifier)
      )
  }) as any
}

export const Entity
  = <Self>() =>
  <Tag extends string, Fields extends Schema.Struct.Fields, MembersOf extends EntitySchema[] = never>(
    tag: Tag, 
    args: {
      fields: Fields,
      membersOf?: MembersOf
    },
    namespace: string
  ): Entity<Self, Tag, Fields, MembersOf> => {
      const membersOf = Schema.optional(args.membersOf ? Schema.Array(Schema.Union(...args.membersOf)) : Schema.Undefined)
      const annotations = namespace ? { [CedarNamespace]: namespace } : undefined
      const base = (TaggedClassImpl<Self>(tag)(tag, { ...args.fields, id: Schema.String, parents: membersOf }, { ...annotations, [EntityTypeId]: EntityTypeId }) as any)

      const serialize = toApiJSON(base)

      class Class extends base {
        [EntityTypeId] = EntityTypeId
        namespace = namespace

        serialize() {
          return serialize(this)
        }
      }

      return Class as any
    }