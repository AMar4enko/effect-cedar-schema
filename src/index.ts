import { Schema, AST } from '@effect/schema'
import { Console, Effect, Match, Option, pipe } from 'effect'
import { TaggedError } from 'effect/Data'
import { IdentifiableEntity } from './types'
import { PropertySignature } from '@effect/schema/Schema'
import { CedarSchema } from './services'
import { getCedarNamespace } from './annotations'

const astMatcher = Match.type<AST.AST>()

const logBoth = { onFailure: Console.error, onSuccess: Console.log }

class UnsupportedSchema extends TaggedError(`UnsupportedSchema`)<{ ast: AST.AST, message?: string }> {
  public toJSON() {
    return {
      message: `Schema is not supported`,
      ast: JSON.stringify(this.ast.toJSON(), null, 2)
    }
  }
}

const compilePropSignatures = (props: AST.PropertySignature[]): Effect.Effect<Record<string, any>, UnsupportedSchema, CedarSchema> => Effect.gen(function* () {
  const schema = yield* CedarSchema

  const compiledProps = Object.fromEntries(props
      .filter((prop) => {
        return String(prop.name) !== `_tag`
      })
      .map(({ name, type, isOptional }) => {
        return [
          String(name),
          compileAttributeType(type, !isOptional)
        ] as const
      })
  )

  return yield* Effect.all(compiledProps).pipe(
    Effect.tapBoth(logBoth)
  )
})

export const compileAttributeType = (ast: AST.AST, required = false): Effect.Effect<Record<string, any>, UnsupportedSchema, CedarSchema>  => Effect.gen(function* () {
  const maybeUseCommonType = <A>(compileType: Effect.Effect<Record<string, any>, UnsupportedSchema, CedarSchema>) => Effect.gen(function* () {
    const id = Option.getOrUndefined(AST.getIdentifierAnnotation(ast))

    const cedarSchema = yield* CedarSchema
    const namespace = yield* getCedarNamespace(ast)
    const namespaceMap = cedarSchema.namespace.get(namespace) 
      ?? cedarSchema.namespace.set(namespace, { actions: new Map(), commonTypes: new Map(), entityTypes: new Map() }).get(namespace)!

    if (!id) {
      return yield* compileType
    }

    return yield* Effect.fromNullable(namespaceMap.commonTypes.get(id))
      .pipe(
        Effect.catchTag(`NoSuchElementException`, () => {
          return compileType
            .pipe(
              Effect.tap(({ required, ...type }) => Effect.sync(() => 
                namespaceMap.commonTypes.set(id, type)
              )),
              Effect.map(() => ({ type: id }))
            )
        })
      )
  })

  const compileType = astMatcher.pipe(
    Match.tags({
      StringKeyword: () => Effect.succeed({
        type: `String`,
        required
      }),
      NumberKeyword: () => Effect.succeed({
        type: `Long`,
        required
      }),
      BooleanKeyword: () => Effect.succeed({
        type: `Boolean`,
        required
      }),
      TypeLiteral: (ast) => compilePropSignatures([...ast.propertySignatures]).pipe(
        Effect.map((attributes) => ({ type: `Record`, attributes }))
      ),
      Transformation: (ast) => {
        if (ast.from._tag === `TupleType`) {
          return compileAttributeType(ast.from.rest[0].type, required).pipe(
            Effect.map(({ required, ...element}) => ({ type: `Set`, element }))
          )
        }
        return compileEntityClass(ast).pipe(
          Effect.map(({ name }) => ({ type: `Entity`, name, required }))
        )
      },
      Union: (ast) => {
        if (ast.types.filter((type) => type._tag === `UndefinedKeyword`).length === 1) {
          return compileAttributeType(ast.types[0], required)
        }
        return Effect.fail(new UnsupportedSchema({ ast }))
      }
    }),
    Match.orElse((ast) => Effect.fail(new UnsupportedSchema({ ast })))
  )(ast)

  return yield* maybeUseCommonType(compileType)
})

export type Action = { principals: readonly IdentifiableEntity[], resources: readonly IdentifiableEntity[], context?: Schema.Schema.Any }

export interface ActionNewable {
  // biome-ignore lint/suspicious/noExplicitAny: <explanation>
  new (props: any, options: { disableValidation?: boolean }): Schema.TaggedRequest<any, any, any, any, any, any, any, any, any>

  ast: AST.AST
  fields: Schema.Struct.Fields
}

export const normalizePropType = (ast: AST.AST | PropertySignature.AST) => {
  switch(ast._tag) {
    case `PropertySignatureDeclaration`:
      return ast.type
    case `PropertySignatureTransformation`:
      return ast.to.type
    default:
      return ast
  }
}

export const compileEntityClass = (ast: AST.Transformation): Effect.Effect<{ name: string, entity: Record<string, any> }, UnsupportedSchema, CedarSchema> => Effect.gen(function* () {
  const id = yield* AST.getIdentifierAnnotation(ast.to).pipe(
    Effect.orElse(() => Effect.fail(new UnsupportedSchema({ ast, message: `Entity class must have an identifier` })))
  )

  const { from } = ast

  if (from._tag !== `TypeLiteral`) {
    return yield* new UnsupportedSchema({ ast: from, message: `Entity compilation failed: entity is not TaggedClass` })
  }

  const cedarSchema = yield* CedarSchema
  const namespace = yield* getCedarNamespace(ast.to)

  console.log(`Compiling entity ${namespace}::${id}`)

  const namespaceMap = cedarSchema.namespace.get(namespace) 
    ?? cedarSchema.namespace.set(namespace, { actions: new Map(), commonTypes: new Map(), entityTypes: new Map() }).get(namespace)!

  const compile = Effect.gen(function* () {
    const membersOf = yield* Effect.fromNullable(from.propertySignatures.find((prop) => prop.name === `parents`)?.type)
      .pipe(
        Effect.tapBoth(logBoth),
        Effect.flatMap(
          astMatcher.pipe(
            Match.tags({
              Union: (unionAst) => {
                const tuple = unionAst.types.filter((type) => type._tag !== `UndefinedKeyword`)
                if (tuple.length === 0) {
                  return Effect.succeed([])
                }
                return Effect.all((tuple[0] as AST.TupleType).rest.map(({ type }) => 
                  astMatcher.pipe(
                    Match.tags({
                      Union: (ast) => 
                        Console.log(`Compiling Union of membersOf`, ast).pipe(
                          Effect.zipRight(
                            Effect.all(ast.types.map(compileEntity))
                          )
                        )
                    }),
                    Match.orElse(compileEntity)
                  )(type)
                ))
              }
            }),
            Match.orElse((ast) => Effect.fail(new UnsupportedSchema({ ast, message: `Can't compile membersOf for ${namespace}::${id}` })))
          )
        ),
        Effect.catchTag(`NoSuchElementException`, () => new UnsupportedSchema({ ast: from, message: `Can't find membersOf property in ${namespace}::${id}` })),
      )

    return yield* compilePropSignatures(from.propertySignatures.filter((prop) => prop.name !== `parents` && prop.name !== `_tag` && prop.name !== `id`)).pipe(
      Effect.map((attributes) => ({
        memberOfTypes: [membersOf].flat(2).map(({ name }) => name),
        shape: {
          type: `Record`,
          attributes
        }
      }))
    )
  })

  const entity = yield* Effect.fromNullable(namespaceMap.entityTypes.get(id))
    .pipe(
      Effect.catchTag(`NoSuchElementException`, () => {
        return compile
          .pipe(
            Effect.tapError(Console.error),
            Effect.tap((entity) => Effect.sync(() => namespaceMap.entityTypes.set(id, entity)))
          )
      })
    )

  return {
    name: [namespace, id].join(`::`),
    entity
  }
})

export const compileEntity: (ast: AST.AST) => Effect.Effect<{ name: string, entity: Record<string, any> }, UnsupportedSchema, CedarSchema> = astMatcher.pipe(
  Match.tags({
    Transformation: (ast) => compileEntityClass(ast),
  }),
  Match.orElse((ast) => Effect.fail(new UnsupportedSchema({ ast })))
)

export const compileAction = <A extends Action>(A: ActionNewable) => Effect.gen(function* () {
  const { principal, resource, context } = A.fields
  const cedarSchema = yield* CedarSchema

  const id = yield* pipe(
    A.ast._tag === `Transformation`
      ? AST.getIdentifierAnnotation(A.ast.to)
      : Option.none(),
    Effect.orElseFail(() => new UnsupportedSchema({ ast: A.ast, message: `Action must have an identifier` }))
  )

  /**
   * In Action schema fields, principals and resources are single-element tuples
   * where element could be either Entity or union of Entities
   */
  const compileEntities = (prop: AST.AST | PropertySignature.AST) => 
    pipe(
      normalizePropType(prop),
      astMatcher.pipe(
        Match.tags({
          Union: (ast) => Effect.all(ast.types.map(compileEntity)),
          Transformation: (ast) => compileEntityClass(ast),
          TupleType: (ast) => {
            return astMatcher.pipe(
              Match.tags({
                Transformation: (ast) => compileEntityClass(ast),
                Union: (ast) => {
                  return Effect.all(ast.types.map(compileEntity))
                },
              }),
              Match.orElse((ast) => Effect.fail(new UnsupportedSchema({ ast })))
            )(ast.rest[0].type)
          }
        }),
        Match.orElse((ast) => Effect.fail(new UnsupportedSchema({ ast })))
      )
    )
    
  const principalsCompiled = yield* compileEntities(principal.ast)
  const resourcesCompiled = yield* compileEntities(resource.ast)

  const contextCompiled = yield* context 
    ? pipe(normalizePropType(context.ast), compileAttributeType)
    : Effect.succeed(undefined)

  const action = {
    appliesTo: {
      principalTypes: [principalsCompiled].flat().map(({ name }) => name),
      resourceTypes: [resourcesCompiled].flat().map(({ name }) => name),
      context: contextCompiled
    }
  }

  const namespaceMap = cedarSchema.namespace.get(cedarSchema.defaultNamespace) 
    ?? cedarSchema.namespace.set(cedarSchema.defaultNamespace, { actions: new Map(), commonTypes: new Map(), entityTypes: new Map() }).get(cedarSchema.defaultNamespace)!
  
  namespaceMap.actions.set(id, action)
}) 

export const compileFromActions = <Actions extends Action[]>(
  actions: {
    [key in keyof Actions]: ActionNewable
  }
) => Effect.gen(function *() {

  yield* Effect.forEach(actions, (A) => compileAction(A as any))

  const { namespace } = yield* CedarSchema

  return Object.fromEntries([...namespace.entries()].map(([key, value]) => [
    key,
    {
      commonTypes: Object.fromEntries(value.commonTypes.entries()),
      entityTypes: Object.fromEntries(value.entityTypes.entries()),
      actions: Object.fromEntries(value.actions.entries())
    }
  ]))

})