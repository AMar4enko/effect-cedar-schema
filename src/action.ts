import { AST, Schema, Serializable } from "@effect/schema"
import { IdentifiableEntity, SerializedType } from "./types"
import { Class, optional, TaggedRequest, TaggedRequestClass } from "@effect/schema/Schema"
import { Effect } from "effect"
import { CedarNamespace } from "./annotations.js"
import { SerializedIdentifier } from "./types.js"
import { compile } from "./entity.js"

const ActionTypeId: unique symbol = Symbol(`effect-cedar/ActionTypeId`)

export const DeterminingPolicies = Schema.Array(Schema.Struct({ policyId: Schema.String }))
export const Errors = Schema.Array(Schema.Struct({ errorDescription: Schema.String }))

export const Success = Schema.Struct({
  DeterminingPolicies: DeterminingPolicies
})

export const Failure = Schema.Struct({
  errors: Errors
})

export type ActionPayload<
  Principals extends Schema.Schema<IdentifiableEntity, any, any>[], 
  Resources extends Schema.Schema<IdentifiableEntity, any, any>[],
  Context extends  Schema.Struct<any> = never
> = [Context] extends [never] ? {
    principal: optional<Schema.Union<Principals>>,
    resource: Schema.Union<Resources>
} : {
    principal: optional<Schema.Union<Principals>>,
    resource: Schema.Union<Resources>,
    context: Context,
}

export interface Action<
  Self, 
  Tag extends string, 
  Principals extends Schema.Schema<IdentifiableEntity, any, any>[], 
  Resources extends Schema.Schema<IdentifiableEntity, any, any>[],
  Context extends Schema.Struct<any> = never
> extends Class<
  Self,
  ActionPayload<Principals, Resources, Context>,
  Schema.Struct.Encoded<ActionPayload<Principals, Resources, Context>>,
  Schema.Struct.Context<ActionPayload<Principals, Resources, Context>>,
  Schema.Struct.Constructor<ActionPayload<Principals, Resources, Context>>,
  TaggedRequest<
    Tag,
    Self,
    Schema.Struct.Encoded<ActionPayload<Principals, Resources, Context>>,
    Schema.Struct.Context<ActionPayload<Principals, Resources, Context>>,
    typeof Success.Type,
    typeof Success.Encoded,
    typeof Failure.Type,
    typeof Failure.Encoded,
    typeof Success.Context | typeof Failure.Context
  >,
  {
    readonly _tag: Tag
    readonly success: typeof Success.Type
    readonly failure: typeof Failure.Type
    readonly namespace: string
    serialize(): Effect.Effect<{
      action: {
        actionId: string
        actionType: string
      },
      principal: SerializedIdentifier,
      resource: SerializedIdentifier,
      context?: {
        contextMap: Record<string, SerializedType>
      }
    }>
  }
> {}

export const serializeAction = (ctx?: AST.AST) => {
  const serializeContext = ctx ? compile(ctx) : Effect.succeed

  return (action: InstanceType<Action<any, any, any, any, any>>) => Effect.gen(function* () {
    let res: Record<string, any> = {}

    if (action.principal) {
      res.principal = yield* (action.principal as any).serialize() 
    }

    if (action.context) {
      res.context = yield* serializeContext(action.context)
    }

    const resource = yield* (action.resource as any).serialize()

    return {
      ...res,
      action: {
        actionType: action.namespace,
        actionId: action._tag
      },
      resource,
    }
  })
}

export const Action
  = <Self>() =>
  <Tag extends string, Principals extends Schema.Schema<any, any, IdentifiableEntity>[], Resources extends Schema.Schema<any, any, IdentifiableEntity>[], Context extends Schema.Struct.Fields = never>(
      tag: Tag,
      args: {
        principals: Principals,
        resources: Resources,
        context?: Context
      },
      namespace: string
    ) => {
      const annotations = namespace ? { [CedarNamespace]: namespace } : undefined
      const payload = {
        principal: Schema.optional(Schema.Union(...args.principals)), 
        resource: Schema.Union(...args.resources),
      }

      const ctx = args.context ? Schema.Struct(args.context) : Schema.Never

      const _serialize = serializeAction(args.context ? ctx.ast : undefined)

      class Cls extends TaggedRequest<Cls>(tag)(tag, {
          success: Success,
          failure: Failure,
          payload: args.context ? {
            ...payload,
            context: ctx
          } : payload
        }, annotations) {
          namespace = namespace
         
          serialize = () => _serialize(this as any)
        }

      Object.defineProperty(Cls, "name", { value: tag });

      return Cls as unknown as Action<Self, Tag, Principals, Resources, [Context] extends [never] ? never : Schema.Struct<Context>>
    }

