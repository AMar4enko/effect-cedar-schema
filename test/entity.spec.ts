import { expect, test } from 'vitest'
import { Author, Book } from './fixtures/entities'
import { Effect } from 'effect'
import { getEntities } from '../src'

test('should be able to serialize', () => {
  const eff = new Book({ title: `My precious book`, id: `123`, ageFrom: 10, author: new Author({ id: `1`, name: `John` }) }).serialize()
    .pipe(
      Effect.bindTo(`value`),
      Effect.bind(`entities`, () => getEntities),
    )

  expect(Effect.runSync(eff)).toMatchObject({
    value: {
      entityType: `BookStore::Book`,
      entityId: `123`
    },
    entities: [
      {
        identifier: {
          entityType: `BookStore::Author`,
          entityId: `1`
        },
        attributes: {
          name: `John`
        }
      },
      {
        identifier: {
          entityType: `BookStore::Book`,
          entityId: `123`
        },
        attributes: {
          title: `My precious book`,
          ageFrom: 10,
          author: {
            entityType: `BookStore::Author`,
            entityId: `1`
          }
        }
      }
    ]
  })
})

