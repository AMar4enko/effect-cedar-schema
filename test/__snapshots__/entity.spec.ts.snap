// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should be able to serialize 1`] = `
{
  "entities": [
    {
      "attributes": {
        "name": {
          "string": "John",
        },
      },
      "identifier": {
        "entityId": "1",
        "entityType": "BookStore::Author",
      },
      "parents": [],
    },
    {
      "attributes": {
        "ageFrom": {
          "long": 10,
        },
        "author": {
          "entityIdentifier": {
            "entityId": "1",
            "entityType": "BookStore::Author",
          },
        },
        "title": {
          "string": "My precious book",
        },
      },
      "identifier": {
        "entityId": "123",
        "entityType": "BookStore::Book",
      },
      "parents": [],
    },
  ],
  "value": {
    "entityId": "123",
    "entityType": "BookStore::Book",
  },
}
`;
